diff --git a/jvector-twenty/src/main/java/io/github/jbellis/jvector/vector/SimdOps.java b/jvector-twenty/src/main/java/io/github/jbellis/jvector/vector/SimdOps.java
index 65fce0b..d185594 100644
--- a/jvector-twenty/src/main/java/io/github/jbellis/jvector/vector/SimdOps.java
+++ b/jvector-twenty/src/main/java/io/github/jbellis/jvector/vector/SimdOps.java
@@ -128,7 +128,6 @@ final class SimdOps {
 
         final int vectorizedLength = FloatVector.SPECIES_64.loopBound(length);
         FloatVector sum = FloatVector.zero(FloatVector.SPECIES_64);
-        sum.selectFrom()
         int i = 0;
         // Process the vectorized part
         for (; i < vectorizedLength; i += FloatVector.SPECIES_64.length()) {
diff --git a/pom.xml b/pom.xml
index 5d17acd..885144c 100644
--- a/pom.xml
+++ b/pom.xml
@@ -120,7 +120,7 @@
                                 <additionalJOption>--add-modules=jdk.incubator.vector</additionalJOption>
                                 <additionalJOption>--enable-preview</additionalJOption>
                             </additionalJOptions>
-                            <release>20</release>
+                            <release>21</release>
                         </configuration>
                     </execution>
                 </executions>
